src/main/java/org/jenkinsci/plugins/KeycloakAccess.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.util.EntityUtils;
+import org.keycloak.adapters.KeycloakDeployment;
+import org.keycloak.authorization.client.AuthzClient;
+import org.keycloak.authorization.client.Configuration;
+import org.keycloak.representations.AccessTokenResponse;
+import org.keycloak.util.JsonSerialization;
+import org.springframework.dao.DataAccessException;
+import org.springframework.dao.DataRetrievalFailureException;
+
+public class KeycloakAccess {
+	private static final Logger LOGGER = Logger.getLogger( KeycloakAccess.class.getName() );
+
+	private KeycloakDeployment keycloakDeployment;
+
+	public KeycloakAccess( KeycloakDeployment keycloakDeployment ) {
+		this.keycloakDeployment = keycloakDeployment;
+	}
+
+	public UserDetails loadUserByUsername( String username ) throws UsernameNotFoundException, DataAccessException {
+		LOGGER.finer( "Requested User Details for: " + username );
+		try {
+			String token = getAuthToken();
+			List<GrantedAuthority> authorities = getAuthorities(getRolesForUser( username, token ));
+			return new KeycloakUserDetails( username, authorities.toArray( new GrantedAuthority[0] ) );
+		} catch ( IOException e ) {
+			LOGGER.log( Level.INFO, "Unable to get user information from Keycloak for: " + username, e );
+			throw new DataRetrievalFailureException( "Unable to get user information from Keycloak for " + username, e );
+		}
+	}
+
+	public GroupDetails loadGroupByGroupname( String groupName ) throws UsernameNotFoundException, DataAccessException {
+		try {
+			List<String> roles = getRoles();
+			if ( roles.contains( groupName ) ) {
+				return new GroupDetails() {
+					@Override
+					public String getName() {
+						return groupName;
+					}
+				};
+			} else {
+				LOGGER.info( "Couldn't find role information for: " + groupName );
+				throw new DataRetrievalFailureException( "Couldn't find role information for: " + groupName );
+			}
+		} catch ( IOException e ) {
+			LOGGER.log( Level.INFO, "Unable to get role information from Keycloak for: " + groupName, e );
+			throw new DataRetrievalFailureException( "Unable to get role information from Keycloak for: " + groupName, e );
+		}
+	}
+
+	public AccessTokenResponse getAccessToken( String username, String password ) {
+		Configuration configuration = new Configuration( keycloakDeployment.getAuthServerBaseUrl(), keycloakDeployment.getRealm(),
+			keycloakDeployment.getResourceName(), keycloakDeployment.getResourceCredentials(), keycloakDeployment.getClient() );
+		AuthzClient authzClient = AuthzClient.create( configuration );
+		return authzClient.obtainAccessToken( username, password );
+	}
+
+	public UserDetails authenticate( String username, String password ) throws AuthenticationException {
+		try {
+			Configuration configuration = new Configuration( keycloakDeployment.getAuthServerBaseUrl(), keycloakDeployment.getRealm(),

```

Comment:
Should we log the successful / unsuccessful authentication of the user here for security purposes?

Reply:
Added successful authentication logging. Unsuccessful logging was already present.